File: samples/worker.py
import threading
import time

from cv2 import waitKey


class WorkerThread(threading.Thread):
    """A worker thread that runs a task at a specified interval."""

    def __init__(self, interval=1):
        super().__init__()
        self.interval = interval
        self.running = True  # Control flag for stopping the thread
        self.daemon = True   # Set as daemon so it stops when the main program exits

    def run(self):
        while self.running:
            print("Thread is running...")
            time.sleep(self.interval)

    def stop(self):
        self.running = False


def main():
    # Create and start the thread
    worker = WorkerThread(interval=2)
    worker.start()

    print("Main function is running...")

    # Main program continues
    while True:
        print("Working from main")
        time.sleep(2)
        key = waitKey(1)
        if key == 27:
            break

    worker.stop()   # Gracefully stop the thread
    worker.join()   # Ensure the thread terminates properly


if __name__ == "__main__":
    main()

File: samples/worker.cpp
/**
 * @file worker.cpp
 * @author Xuhua Huang
 * @brief
 *
 * g++ -c worker.cpp -o worker.exe -std=c++2a -Wall -Wextra -O3
 *
 * @version 0.1
 * @date 2025-02-25
 *
 * @copyright Copyright (c) 2025
 *
 */

#include <atomic>
#include <chrono>
#include <iostream>
#include <thread>
// #include <opencv2/opencv.hpp>

class WorkerThread {
public:
    WorkerThread(int interval)
        : interval(interval)
        , running(true) {}

    void start() { workerThread = std::thread(&WorkerThread::run, this); }

    void stop() {
        running = false;
        if (workerThread.joinable()) {
            workerThread.join();
        }
    }

    ~WorkerThread() { stop(); }

private:
    int               interval;
    std::atomic<bool> running;
    std::thread       workerThread;

    void run() {
        while (running) {
            std::cout << "Thread is running..." << "\n";
            std::this_thread::sleep_for(std::chrono::seconds(interval));
        }
    }
};

int main() {
    WorkerThread worker(2);
    worker.start();

    while (true) {
        std::cout << "Working from main" << "\n";

        // if (cv::waitKey(1) == 27) { // ESC key
        if (std::cin.get() == 27) {
            std::cout << "ESC key pressed" << "\n";
            break;
        }
    }

    worker.stop();
    return 0;
}
